### PR TITLE
Diffs, Results, and Dry Runs

### PR DESCRIPTION
**Pull request by rsandor**
This adds the following major features and fixes to `fs-transform`:

1. `getDiff()` - Allows the user to get a full text diff of what changed after performing transformations.
2. `.dry()` - Allows the user to perform a dry run without changing the given root directory.
3. `.results` - Keeps track of all diffs, commands, and renames that occurred for each transformation rule executed by the `Transformer`.
4. Shell commands now work correctly in both OS X and Linux

cc: @Runnable/owners 
It's big, please take a little time and make sure to ask a lot of questions when reviewing.

### CODE REVIEW

**Comment by podviaznikov** [comment]
hm, reviewed code once and have no comments. Will try do second round later today

File: package.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
   "description": "Fast, rule based, file system transformations.",
   "main": "index.js",
   "scripts": {
-    "test": "jshint index.js && lab -v -c test/*",
+    "test": "jshint index.js lib/ && lab -v -c -a code test/*",
```
Comment (bkendall) [inline package.json]: :+1:

**Comment by bkendall** [comment]
looks good to me - I guess you'll really find out how you did when you start using this more ;)

**Comment by podviaznikov** [comment]
+1. agree with @bkendall

File: lib/fs-driver.js [hunk 2]
```diff
@@ -98,13 +139,11 @@ FsDriver.prototype.move = function (source, dest, cb) {
  * @return {string} The copy command to be executed.
  */
 FsDriver.prototype.copy = function (source, dest, cb) {
-  var command = [
+  return this.exec([
```
Comment (anandkumarpatel) [inline lib/fs-driver.js]: having a callback and a return value might get confusing after a while. is there a reason you don't put the return value in the `cb` ?

File: lib/fs-driver.js [hunk 3]
```diff
@@ -173,3 +206,100 @@ FsDriver.prototype.sed = function (search, replace, name, line, cb) {
...
+  ].join(' '), cb);
+};
+
+/**
+ * Attempts to find an available working directory relative to /tmp.
+ */
+FsDriver.prototype.findWorkingDirectory = function() {
+  var rootName = last(this.root.split('/'));
+  var workingPath = '/tmp/.' + rootName + '.fs-work';
+  var n = 0;
```
Comment (anandkumarpatel) [inline lib/fs-driver.js]: what if you used `Math.random();` here? should reduce time to generate a working dir

**Comment by anandkumarpatel** [comment]
LGTM, one thing we should do though. 
since you are using external tools (sed, move, diff) on init you should check that those exist and are in the correct version.

also add this check to the output script. people using minimalistic docker images so they might not have these programs.

File: lib/fs-driver.js (hunk 3 above)
Comment (bkendall) [inline lib/fs-driver.js]: +1 - while loops can be bad - while it's unlikely that many exist, using a random something generator may be a better move

**Comment by rsandor** [comment]
@anandkumarpatel - I agree, we should check if the tools exist. Doing version checking would be tricky for multiple OSes though. Considering that these are pretty well distributed and common tools I think it's safe to assume that if it is installed, it's probably the proper version. Thoughts?

**Comment by anandkumarpatel** [comment]
*safe to assume* famous last words :wink: . its going to be a long task to figure out which versions work/done with for this, we can hold off until a user hits these obscure bugs (will be 1 in a million probably) 

however I want to push for an existence test since a lot of docker images do not have these tools installed.

**Comment by rsandor** [comment]
@anandkumarpatel Check this out: https://github.com/Runnable/fs-transform/commit/6db0582f69a38fb62f007f9103aaea88858b2f12

**Comment by anandkumarpatel** [comment]
:+1: LGTM! nice work

**Pull request merged**